### PR TITLE
(KEP-2436) Task: Controller Manager Leader Migration

### PR DESCRIPTION
**Pull request by jiahuif**
This PR creates an example use case of Leader Migration, a new feature defined in KEP-2436 coming alpha in the v1.21 release. Leader Migration can transfer the ownership of controllers across controller managers while keeping the controllers available and only running under one controller manager in a HA setting.

The task shows all necessary steps to migrate a control plane from `kube-controller-manager` only to `kube-controller-manager` + `cloud-controller-manager`, which will be necessary for upgrading any cluster that uses soon-to-be-gone in-tree cloud-providers. A summary of steps is as follows.
 
1. Grant necessary permission for the migration lease
2. Create the "before" configuration file for the old KCM
3. Create the "after" configuration file for new CCM
4. Upgrade the cluster from the "before" state to "after" state
5. Optionally, disable Leader Migration

Because the documentation is more user-facing, implementation details are omitted. Please refer to the KEP if interested.

References:
KEP: https://github.com/kubernetes/enhancements/issues/2436

### CODE REVIEW

**Comment by annajung** [comment]
cc @reylejano

**Comment by jiahuif** [comment]
Switching base from `master` to `dev-1.21`

**Comment by reylejano** [comment]
Hi @jiahuif, Doc PRs for the 1.21 release should be Ready for Review by March 26. When this PR is ready for review, please change the PR title to remove the words wip or placeholder when PR is ready for review (or change the stage from draft PR to ready for review)

Upcoming doc related dates for the 1.21 release:
- Docs Ready for Review deadline is March 24
- Docs Ready to Merge deadline is March 31

**Comment by reylejano** [comment]
@jiahuif @jpbetz @cheftako The 1.21 Docs Ready for Review deadline is tomorrow. If this PR is ready for review, please retitle this PR and remove [WIP]

Thank you

**Comment by jiahuif** [comment]
Finished the `reviewers` metadata, and it is good for reivew.

File: content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+
+## {{% heading "prerequisites" %}}
+
+The control plane nodes should run `kube-controller-manager` with Leader Election enabled through `--leader-elect=true`, and the cloud privider set with `--cloud-provider` flag. At this moment, `cloud-controller-manager` should not have been deployed.
+
+<!-- steps -->
+
+### Grant Access to Migration Lease
+
+The default premissions of controller manager allow only access to their main lease. In order for the migration to work, access to another lease is required.
```
Comment (tengqm) [inline content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md:33]:
```suggestion
The default permissions of the controller manager allow only accesses to their main lease. In order for the migration to work, accesses to another lease are required.
```

File: content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+We can grant full access of leases API to `kube-controller-manager` by modifying `system::leader-locking-kube-controller-manager` role. In this guide, we assume the name of the migration lease is `cloud-provider-extraction-migration`.
```
Comment (tengqm) [inline content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md:35]:
```suggestion
We can grant `kube-controller-manager` full access to the leases API by modifying the `system::leader-locking-kube-controller-manager` role. In this guide, we assume the name of the migration lease is `cloud-provider-extraction-migration`.
```

Comment (tengqm) [inline content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md:35]:
```suggestion
We can grant `kube-controller-manager` full access to the leases API by modifying the `system::leader-locking-kube-controller-manager` role. In this guide, we assume the name of the migration lease is `cloud-provider-extraction-migration`.
```

File: content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+`kubectl patch -n kube-system role 'system::leader-locking-kube-controller-manager' -p '{"rules": [ {"apiGroups":[ "coordination.k8s.io"], "resources": ["leases"], "resourceNames": ["cloud-provider-extraction-migration"], "verbs": ["create", "list", "get", "update"] } ]}' --type=merge`
+
+Do the same to `system::leader-locking-cloud-controller-manager` role.
```
Comment (tengqm) [inline content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md:39]:
```suggestion
Do the same to the `system::leader-locking-cloud-controller-manager` role.
```

File: content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+  - name: service
+    component: kube-controller-manager
+  - name: cloud-node-lifecycle
+    component: kube-controller-manager
+
+```
+
+On each control plane node, save the content to `/etc/leadermigration.conf`, and update the manifest of `kube-controller-manager` so that the file is mounted inside the container at the same location. Also, update the same manifest to add the following arguments:
+
+- `--feature-gates=ControllerManagerLeaderMigration=true` to enable Leader Migration as an alpha feature
```
Comment (tengqm) [inline content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md:64]:
```suggestion
- `--feature-gates=ControllerManagerLeaderMigration=true` to enable Leader Migration which is an alpha feature
```

**Comment by reylejano** [comment]
@jpbetz @cheftako , can you provide a technical review for this PR by March 31 to get this into the release. Thank you!

File: content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+  - name: route
+    component: kube-controller-manager
+  - name: service
+    component: kube-controller-manager
+  - name: cloud-node-lifecycle
+    component: kube-controller-manager
+
+```
+
+On each control plane node, save the content to `/etc/leadermigration.conf`, and update the manifest of `kube-controller-manager` so that the file is mounted inside the container at the same location. Also, update the same manifest to add the following arguments:
```
Comment (jpbetz) [inline content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md:62]: "mounted inside the container at the same location"

Probably worth showing this in yaml as well-- it saves a bit of time and it minimizes changes of someone misunderstanding the step.

File: content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+
+### Upgrade Control Plane
+
+The control plane now contains nodes of both versions. The old version running `kube-controller-manager` only, and the new one running both `kube-controller-manager` and `cloud-controller-manager`. The migrated controllers, as specified in the configuration, are running under either old `kube-controller-manager` or `cloud-controller-manager` but not both at any time, depending on which controller manager holds the migration lease.
+
+In a rolling manner, create a new control plane node and bring down one of the old until the control plane contains only nodes of the new version.
+
+### (Optional) Disable Leader Migration
+
+Now that the control plane has been upgraded to run both `kube-controller-manager` and `cloud-controller-manager`, Leader Migration has finished its job and can be safely disabled to save one lease resource.
```
Comment (jpbetz) [inline content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md:103]: It might be needed again if a rollback needs to happen. The rollback might be required for some completely unrelated problem. Maybe recommend that cluster administrators wait an additional release before doing this?

File: content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+
+{{< glossary_definition term_id="cloud-controller-manager" length="all" prepend="The cloud-controller-manager is">}}
+
+## Background
+
+Kubernetes project is in the process of migrating the cloud specific code from the core Kubernetes tree to external packages and removing them from the `kube-controller-manager`. Once the extraction has been completed, existing users running older versions of Kubernetes need a process to migrate their existing clusters to use the new `cloud-controller-manager` component with minimal risk.
+
+Leader Migration provides a mechanism in which HA clusters can safely migrate "cloud specific" controllers between the `kube-controller-manager` and the `cloud-controller-manager` via a shared resource lock between the two components.
+
+This guide walks you through the process of upgrading the control plane from `kube-controller-manager` with built-in cloud provider to running both `kube-controller-manager` and `cloud-controller-manager`.
```
Comment (jpbetz) [inline content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md:23]: Point out that these are the manual steps required to do the migration, but that cluster administration tools and cloud provider specific documentation might offer more details or an automated workflow?

File: content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md (hunk 5 above)
Comment (jpbetz) [reply content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md:62]: Oh, also, not all cluster configurations use the term "manifest". Might be clearer to say call it something like the "Pod declaration for the `kube-controller-manager`"?

File: content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+  - name: route
+    component: cloud-controller-manager
+  - name: service
+    component: cloud-controller-manager
+  - name: cloud-node-lifecycle
+    component: cloud-controller-manager
+
+```
+
+When creating new control plane nodes, the content should be deploy to `/etc/leadermigration.conf`. The manifest of `cloud-controller-manager` should be updated to mount the configuration file in the same manner as `kube-controller-manager`. Similarly, add `--feature-gates=ControllerManagerLeaderMigration=true`,`--enable-leader-migration`, and `--leader-migration-config=/etc/leadermigration.conf` to the arguments.
```
Comment (jpbetz) [inline content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md:89]: Is there a convention to use the ".conf" file type for configuration yaml? I was expecting ".yaml" but I might be wrong.

File: content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+
+<!-- overview -->
+
+{{< feature-state state="alpha" for_k8s_version="v1.21" >}}
+
+{{< glossary_definition term_id="cloud-controller-manager" length="all" prepend="The cloud-controller-manager is">}}
+
+## Background
+
+Kubernetes project is in the process of migrating the cloud specific code from the core Kubernetes tree to external packages and removing them from the `kube-controller-manager`. Once the extraction has been completed, existing users running older versions of Kubernetes need a process to migrate their existing clusters to use the new `cloud-controller-manager` component with minimal risk.
```
Comment (jpbetz) [inline content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md:19]:
```suggestion
As part of the cloud provider extraction effort, all cloud specific controllers must be moved out of the `kube-controller-manager`. All existing clusters that run cloud controllers in the `kube-controller-manager` must migrate to instead run the controllers in a cloud provider specific `cloud-controller-manager`.
```

**Review by jpbetz — CHANGES_REQUESTED** [review]

**Comment by jpbetz** [comment]
> Joe Betz Walter Fender , can you provide a technical review for this PR by March 31 to get this into the release. Thank you!

Thanks @reylejano, I've done a first pass and will be sure to get all technical review done by the 31st.

File: content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+
+Please refer to [Cloud Controller Manager Administration](/docs/tasks/administer-cluster/running-cloud-controller/) for more detail on how to deploy `cloud-controller-manager`.
+
+### Upgrade Control Plane
+
+The control plane now contains nodes of both versions. The old version running `kube-controller-manager` only, and the new one running both `kube-controller-manager` and `cloud-controller-manager`. The migrated controllers, as specified in the configuration, are running under either old `kube-controller-manager` or `cloud-controller-manager` but not both at any time, depending on which controller manager holds the migration lease.
+
+In a rolling manner, create a new control plane node and bring down one of the old until the control plane contains only nodes of the new version.
+
+### (Optional) Disable Leader Migration
```
Comment (sftim) [inline content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md:101]: Let's leave the “optional” out of the fragment identifier:
```suggestion
### (Optional) Disable Leader Migration {#disable-leader-migration}
```

File: content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+
+Now that the control plane has been upgraded to run both `kube-controller-manager` and `cloud-controller-manager`, Leader Migration has finished its job and can be safely disabled to save one lease resource.
```
Comment (sftim) [inline content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md:103]:
```suggestion
Now that the control plane has been upgraded to run both `kube-controller-manager` and `cloud-controller-manager`, Leader Migration has finished its job and can be safely disabled to save one Lease resource.
```
?

File: content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+## Background
+
+As part of the cloud provider extraction effort, all cloud specific controllers must be moved out of the `kube-controller-manager`. All existing clusters that run cloud controllers in the `kube-controller-manager` must migrate to instead run the controllers in a cloud provider specific `cloud-controller-manager`.
+
+Leader Migration provides a mechanism in which HA clusters can safely migrate "cloud specific" controllers between the `kube-controller-manager` and the `cloud-controller-manager` via a shared resource lock between the two components.
+
+This guide walks you through the process of upgrading the control plane from `kube-controller-manager` with built-in cloud provider to running both `kube-controller-manager` and `cloud-controller-manager`.
+
+## {{% heading "prerequisites" %}}
+
```
Comment (sftim) [inline content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md:26]: AIUI, readers need
- a cluster
- that is already integrated with an in-tree cloud provider

File: content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+The control plane nodes should run `kube-controller-manager` with Leader Election enabled through `--leader-elect=true`, and the cloud privider set with `--cloud-provider` flag.`cloud-controller-manager` should not yet be deployed.
+
+<!-- steps -->
+
+### Grant Access to Migration Lease
```
Comment (sftim) [inline content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md:31]: nit: headings should be sentence case (but _page_ titles are title case).

File: content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+
+The default permissions of the controller manager allow only accesses to their main lease. In order for the migration to work, accesses to another lease are required.
+
+We can grant `kube-controller-manager` full access to the leases API by modifying the `system::leader-locking-kube-controller-manager` role. In this guide, we assume the name of the migration lease is `cloud-provider-extraction-migration`.
```
Comment (sftim) [inline content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md:35]: nit:
```suggestion
You can grant `kube-controller-manager` full access to the leases API by modifying the `system::leader-locking-kube-controller-manager` role. This task guide assumes that the name of the migration lease is `cloud-provider-extraction-migration`.
```
(see [avoid using we](https://kubernetes.io/docs/contribute/style/style-guide/#avoid-using-we))

File: content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md [hunk 15]
```diff
@@ -0,0 +1,109 @@
+---
+reviewers:
+- jpbetz
+- cheftako
+title: "Controller Manager Leader Migration"
+linkTitle: "Controller Manager Leader Migration"
```
Comment (sftim) [inline content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md:6]: Possibly “Migrate A Replicated Control Plane To Use The cloud-controller-manager”?

File: content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md [hunk 16]
```diff
@@ -0,0 +1,109 @@
...
+
+<!-- overview -->
+
+{{< feature-state state="alpha" for_k8s_version="v1.21" >}}
+
+{{< glossary_definition term_id="cloud-controller-manager" length="all" prepend="The cloud-controller-manager is">}}
+
+## Background
+
+As part of the cloud provider extraction effort, all cloud specific controllers must be moved out of the `kube-controller-manager`. All existing clusters that run cloud controllers in the `kube-controller-manager` must migrate to instead run the controllers in a cloud provider specific `cloud-controller-manager`.
```
Comment (sftim) [inline content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md:19]: > cloud provider extraction

Can we link to more details about this, here / later in the page?

**Review by sftim — COMMENTED** [review]
Thanks @jiahuif 

Here's some feedback - I hope it's useful

Anything marked as a nit, it's OK to leave in (we can merge it even if that's left the way it is now). Anything that's a question _is a question_ - you don't have to make suggested changes!

File: content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md (hunk 5 above)
Comment (jiahuif) [reply content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md:62]: Term "manifest" is defined by kubelet, as "Manifests are standard Pod definitions in JSON or YAML format in a specific directory." Ref: https://kubernetes.io/docs/tasks/configure-pod-container/static-pod/#configuration-files

Because for a control plane node, pods like kube-api-server, kube-controller-manager, etc, are started as static pods by kubelet, I think it is ok to use "manifest" here.

File: content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md (hunk 8 above)
Comment (jiahuif) [reply content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md:89]: There is no defined convention. However, we have kubeconfig and credentials save as `.conf`, so I followed them.

File: content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md (hunk 13 above)
Comment (jiahuif) [reply content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md:31]: Thank you. Updating.

File: content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md (hunk 12 above)
Comment (jiahuif) [reply content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md:26]: That is a good point. Will update prerequisites section.

File: content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md (hunk 5 above)
Comment (sftim) [reply content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md:62]: Maybe mention “static pod” to make it clear we're assuming that the control plane components are run that way. The manifests for static pods are, indeed, manifests.

File: content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md [hunk 17]
```diff
@@ -0,0 +1,109 @@
...
+
+## {{% heading "prerequisites" %}}
+
+The control plane nodes should run `kube-controller-manager` with Leader Election enabled through `--leader-elect=true`, and the cloud privider set with `--cloud-provider` flag.`cloud-controller-manager` should not yet be deployed.
+
+<!-- steps -->
+
+### Grant Access to Migration Lease
+
+The default permissions of the controller manager allow only accesses to their main lease. In order for the migration to work, accesses to another lease are required.
```
Comment (sftim) [inline content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md:33]:
```suggestion
The default permissions of the controller manager allow only accesses to their main Lease. In order for the migration to work, accesses to another Lease are required.
```

File: content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md (hunk 12 above)
Comment (sftim) [inline content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md:26]: BTW it also looks like this guide assumes you're using RBAC authorization.

**Comment by sftim** [comment]
@jiahuif if you could very briefly outline this change in the PR description, that'd help.

The [preview](https://605cddea52e56b0007424246--kubernetes-io-vnext-staging.netlify.app/docs/tasks/administer-cluster/controller-manager-leader-migration/) looks OK. I recommend another round of technical review.

How did you feel about the different task title I suggested?

**Comment by jiahuif** [comment]
@sftim Thanks for suggesting the title change. I agree it will be good if we can use a title that more clearly shows the purpose. I will think of a title based on your suggestion shortly.

File: content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md (hunk 15 above)
Comment (jiahuif) [reply content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md:6]: I simplified it a bit to "Migrate Replicated Control Plane To Use Cloud Controller Manager", matching titles of existing pages. What do you think?

File: content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md (hunk 5 above)
Comment (jiahuif) [reply content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md:62]: Added to prerequisites.

File: content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md (hunk 16 above)
Comment (jiahuif) [reply content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md:19]: Added a link to https://kubernetes.io/blog/2019/04/17/the-future-of-cloud-providers-in-kubernetes/

File: content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md (hunk 6 above)
Comment (jiahuif) [reply content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md]: I think "an additional release" part heavily depends on the out-of-tree cloud provider. Probably better let the user decide.

File: content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md (hunk 12 above)
Comment (jiahuif) [reply content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md:26]: added to prerequisites

File: content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md (hunk 7 above)
Comment (jiahuif) [reply content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md]: clarified this is the manual process.

**Comment by jiahuif** [comment]
Note: squash the commits after the second technical review.

File: content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md (hunk 6 above)
Comment (jiahuif) [reply content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md]: Still, I think it would be great to note that a rollback may happen, and Leader Migration can be enabled back any time as the user likes.

**Comment by reylejano** [comment]
> > Joe Betz Walter Fender , can you provide a technical review for this PR by March 31 to get this into the release. Thank you!
> 
> Thanks @reylejano, I've done a first pass and will be sure to get all technical review done by the 31st.

Hi @jpbetz , can you take a second pass on this PR and give a tech lgtm if the changes and updates are acceptable

**Comment by reylejano** [comment]
@jiahuif please address [sftim's comment](https://github.com/kubernetes/website/pull/26970#issuecomment-807844813) about providing a brief outline of this PR in the PR's description

**Comment by jiahuif** [comment]
The outline as been added. Thank you for the reminder.

File: content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md [hunk 18]
```diff
@@ -0,0 +1,115 @@
...
+
+`kubectl patch -n kube-system role 'system::leader-locking-kube-controller-manager' -p '{"rules": [ {"apiGroups":[ "coordination.k8s.io"], "resources": ["leases"], "resourceNames": ["cloud-provider-extraction-migration"], "verbs": ["create", "list", "get", "update"] } ]}' --type=merge`
+
+Do the same to the `system::leader-locking-cloud-controller-manager` role.
+
+`kubectl patch -n kube-system role 'system::leader-locking-cloud-controller-manager' -p '{"rules": [ {"apiGroups":[ "coordination.k8s.io"], "resources": ["leases"], "resourceNames": ["cloud-provider-extraction-migration"], "verbs": ["create", "list", "get", "update"] } ]}' --type=merge`
+
+### Initial Leader Migration configuration
+
+Leader Migration requires a configuration file representing the state of controller-to-manager assignment. The following example configuration shows the assignment at this moment where `route`, `service`, and `cloud-node-lifecycle` controllers are started within `kube-controller-manager`.
```
Comment (cheftako) [inline content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md:51]: Worth mentioning this is the default/normal set of controllers to migrate.

File: content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md [hunk 19]
```diff
@@ -0,0 +1,115 @@
...
+
+```
+
+On each control plane node, save the content to `/etc/leadermigration.conf`, and update the manifest of `kube-controller-manager` so that the file is mounted inside the container at the same location. Also, update the same manifest to add the following arguments:
+
+- `--feature-gates=ControllerManagerLeaderMigration=true` to enable Leader Migration which is an alpha feature
+- `--enable-leader-migration` to enable Leader Migration on the controller manager
+- `--leader-migration-config=/etc/leadermigration.conf` to set configuration file
+
+Restart `kube-controller-manager` on each node. At this moment, `kube-controller-manager` has leader migration enabled and is ready for the migration.
```
Comment (cheftako) [inline content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md:74]: Ideally this is done as a migration from version N (eg v1.21) to version N + 1 (eg v1.22). This should then be always be the KCM configuration at version N. This is important if you end up needing to rollback from N+1 to N. The migration work both upgrading and downgrading assuming it is configured this way.

File: content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md [hunk 20]
```diff
@@ -0,0 +1,115 @@
...
+    component: cloud-controller-manager
+  - name: cloud-node-lifecycle
+    component: cloud-controller-manager
+
+```
+
+When creating new control plane nodes, the content should be deploy to `/etc/leadermigration.conf`. The manifest of `cloud-controller-manager` should be updated to mount the configuration file in the same manner as `kube-controller-manager`. Similarly, add `--feature-gates=ControllerManagerLeaderMigration=true`,`--enable-leader-migration`, and `--leader-migration-config=/etc/leadermigration.conf` to the arguments.
+
+Create a new control plane node with the updated `cloud-controller-manager` manifest, and with the `--cloud-provider` flag unset for `kube-controller-manager`. The newly deployed `kube-controller-manager` MUST NOT have Leader Migration enabled because it does not run the migrated controllers with an external cloud provider and thus is not involved in the migration.
+
```
Comment (cheftako) [inline content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md:98]: This would then be the version N+1 configuration.

File: content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md (hunk 19 above)
Comment (cheftako) [reply content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md:74]: Worth mentioning the KCM should *NOT* be configured this way at version N+1.

Comment (jiahuif) [reply content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md:74]: Will need to add "upgrade from N to N + 1" as an assumption to prerequisites and mention that N to N w/ CCM works too.

**Review by sftim — COMMENTED** [review]
As this is documenting an alpha feature, it's OK to have a few documentation rough edges.

We do want:
- factual accuracy, no obvious incorrect statements etc
- a clear indication that the feature is alpha (DONE!)
- enough detail that someone can decide whether or not that feature is relevant; if they find it's on in their cluster, should they worry, etc.
- accurate feature gate & related details

I think that's the minimum. Obviously going beyond that is better and welcome.

We will consider PRs against _dev-1.21_ even after the initial docs merge, and because this is website is continuously delivered, we'll also consider the same commits post-release if you rebase against the primary branch.

TL;DR: So we don't need this PR to be completely finished feature docs, so long as the minimum is done well.

**Comment by jiahuif** [comment]
- [ ] factual accuracy, no obvious incorrect statements etc (pending review)
- [X] a clear indication that the feature is alpha (DONE!)
- [X] enough detail that someone can decide whether or not that feature is relevant; if they find it's on in their cluster, should they worry, etc.
- [X] accurate feature gate & related details

File: content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md [hunk 21]
```diff
@@ -0,0 +1,115 @@
...
+{{< feature-state state="alpha" for_k8s_version="v1.21" >}}
+
+{{< glossary_definition term_id="cloud-controller-manager" length="all" prepend="The cloud-controller-manager is">}}
+
+## Background
+
+As part of the [cloud provider extraction effort](https://kubernetes.io/blog/2019/04/17/the-future-of-cloud-providers-in-kubernetes/), all cloud specific controllers must be moved out of the `kube-controller-manager`. All existing clusters that run cloud controllers in the `kube-controller-manager` must migrate to instead run the controllers in a cloud provider specific `cloud-controller-manager`.
+
+Leader Migration provides a mechanism in which HA clusters can safely migrate "cloud specific" controllers between the `kube-controller-manager` and the `cloud-controller-manager` via a shared resource lock between the two components.
```
[the remark this answers is not in the document]
Comment (jiahuif) [reply content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md]: Add the cases where we do not need Leader Migration to "Background"

File: content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md (hunk 18 above)
Comment (jiahuif) [reply content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md:53]: Need to add `As of version 1.21` because we may change that in future release

Comment (sftim) [reply content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md:53]: @jiahuif can you explain what you mean? Is this example something we'd need to update for future releases?

File: content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md (hunk 19 above)
Comment (jiahuif) [reply content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md:76]: "KCM should NOT be configured this way at version N+1" was mentioned below.

File: content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md (hunk 18 above)
Comment (jiahuif) [reply content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md]: This three controllers will eventually be removed from `kube-controller-manager` once legacy cloud providers are removed from k/k. Using legacy cloud providers is deprecated, which is exactly why we need to migrate them to cloud-controller-manager.

File: content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md (hunk 19 above)
Comment (jiahuif) [reply content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md:76]: Any reason why it is a bad idea to migrate from N w/ KCM to N w/ KCM+CCM?

File: content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md (hunk 20 above)
Comment (jiahuif) [reply content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md:102]: clarified.

File: content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md (hunk 18 above)
Comment (sftim) [reply content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md]: If we change the set of legacy controllers that need migrating, I'd expect SIG Cloud Provider to update this example as part of documenting the change. The “version 1.21” bit is built into the documentation (we don't mention on each page that the documentation relates to the docs version you're viewing).

Does that sound right / make sense?

**Comment by sftim** [comment]
If you have time, also edit https://kubernetes.io/docs/concepts/architecture/cloud-controller/#what-s-next to link to the new page

**Comment by sftim** [comment]
[Preview](https://60636ae64ad7da00081514a1--kubernetes-io-vnext-staging.netlify.app/docs/tasks/administer-cluster/controller-manager-leader-migration/) of new page

File: content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md (hunk 18 above)
Comment (jiahuif) [reply content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md]: You're right. I think `At this point` should be good enough.

File: content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md [hunk 22]
```diff
@@ -0,0 +1,119 @@
...
+
+In a rolling manner, create a new control plane node of version N + 1 and bring down one of version N + 1 until the control plane contains only nodes of version N + 1.
+
+### (Optional) Disable Leader Migration {#disable-leader-migration}
+
+Now that the control plane has been upgraded to run both `kube-controller-manager` and `cloud-controller-manager` of version N + 1, Leader Migration has finished its job and can be safely disabled to save one Lease resource. If a rollback must be performed later, you can safely re-enable Leader Migration before the rollback.
+
+In a rolling manager, update manifest of `cloud-controller-manager` to unset both `--enable-leader-migration` and `--leader-migration-config=` flag, also remove the mount of `/etc/leadermigration.conf`, and finally remove `/etc/leadermigration.conf`. If you need to re-enable Leader Migration, you can recreate the configuration file and add its mount and the flags that enable Leader Migration back to `cloud-controller-manager`.
+
+## Additional Reading
```
Comment (sftim) [inline content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md:117]: nit: we have a shortcode for an end-of-page “What's next” heading.
```suggestion
## {{% heading "whatsnext" %}}
```

File: content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md [hunk 23]
```diff
@@ -0,0 +1,119 @@
...
+
+[Controller Manager Leader Migration - KEP](https://github.com/kubernetes/enhancements/tree/master/keps/sig-cloud-provider/2436-controller-manager-leader-migration)
```
Comment (sftim) [inline content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md:119]:
```suggestion
- Read the [Controller Manager Leader Migration](https://github.com/kubernetes/enhancements/tree/master/keps/sig-cloud-provider/2436-controller-manager-leader-migration) enhancement proposal
```

**Comment by jiahuif** [comment]
Link to this page added to concept of "cloud controller mangaer"

File: content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md (hunk 19 above)
Comment (cheftako) [reply content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md:77]: Depends on your tooling. It can be easier to track and automate things like upgrade/downgrade if its tied to a release. For example if you migrate at version N then need to downgrade for unrelated reasons, do you end up reverting to KCM only while not using the controller migration lock?

**Comment by cheftako** [comment]
:+1: Seems technically correct.

File: content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md (hunk 19 above)
Comment (jiahuif) [reply content/en/docs/tasks/administer-cluster/controller-manager-leader-migration.md:77]: Stated the reason to be "so that configuration can be aligned to releases"

**Comment by reylejano** [comment]
Thank you all for the reviews and thank you @jiahuif for addressing comments.
/lgtm
/approve
